daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1310,9 +1326,16 @@ private[lf] object SBuiltin {
       methodName: MethodName,
   ) extends SBuiltin(1) {
     override private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
-      val (tyCon, record) = getSAnyContract(args, 0)
-      machine.ctrl =
-        SEApp(SEVal(ImplementsMethodDefRef(tyCon, ifaceId, methodName)), Array(SEValue(record)))
+      val (templateId, record) = getSAnyContract(args, 0)
+      val ref = getImplementsOrCoImplements(machine, templateId, ifaceId) match {
+        case Some(TemplateOrInterface.Template(ImplementsDefRef(_, _))) =>
+          ImplementsMethodDefRef(templateId, ifaceId, methodName)
+        case Some(TemplateOrInterface.Interface(CoImplementsDefRef(_, _))) =>
+          CoImplementsMethodDefRef(templateId, ifaceId, methodName)
+        case None =>
+          throw SErrorDamlException(IE.TemplateDoesNotImplementInterface(ifaceId, templateId))
```

Comment:
I'm not happy with this case. To end up here, a user would need to construct a value of type `ifaceId` from an underlying `templateId`, without `templateId` actually implementing (or being co-implemented by) `ifaceId`, which I don't think is possible. 

The other thing I don't like is that other functions in this module throw `ContractDoesNotImplementInterface`, which takes a `ContractId`, but here we don't have one so we need a different type.

Perhaps if we're sure that this case is unreachable we can just use `SErrorCrash` instead of a new error constructor?

Reply:
oh looks like we got to the same conclusion sorry missed your comment.

Reply:
ah great I wanted to be sure before making it a crash. I'll do it now, thank you!